extends CharacterBody2D
class_name Player

@onready var animator = $AnimationPlayer
@onready var cshape = $CollisionShape2D

var speed = 300.0
var accelerometer_speed = 130.0

var gravity = 15.0
var max_fall_velocity = 1000.0
var jump_velocity = -800

var viewport_size
# ğŸ‘‡ for detecting the platform is either Mobile or PC
var use_accelerometer = false
# ğŸ‘‡ If true, the game will stop taking input for the player
var dead = false
# ğŸ‘‡ To signal to main script -> game script -> game over screen
signal died

# ğŸ‘‡ Turning animation names into variables so that they can
# ğŸ‘‡ be later inside use_new_skin() function
var fall_anim_name = "fall"
var jump_anim_name = "jump"

# ğŸ‘‡ Getting reference of Sprite2D node
@onready var sprite = $Sprite2D

func _ready() -> void:
	viewport_size = get_viewport_rect().size
	
	var os_name = OS.get_name()
	if os_name == "Android" || os_name == "iOS":
		use_accelerometer = true

func _process(_delta: float) -> void:
	if velocity.y > 0:
		if animator.current_animation != fall_anim_name:
			animator.play(fall_anim_name)
	elif velocity.y < 0:
		if animator.current_animation != jump_anim_name:
			animator.play(jump_anim_name)

func _physics_process(_delta: float) -> void:
	velocity.y += gravity
	if velocity.y > max_fall_velocity:
		velocity.y = max_fall_velocity
	
	# ğŸ‘‡ If the player is still alive
	if !dead:
		if use_accelerometer:
			var mobile_input = Input.get_accelerometer()
			# print("Accelerometer: " + str(mobile_input))
			# ğŸ‘‡ The value of "mobile_input.x" swings bwetween -10 and +10 for left to right direction.
			# ğŸ‘‡ "accelerometer_speed" is multiplied to give a decent speed
			velocity.x = mobile_input.x * accelerometer_speed
		else:
		# ğŸ‘‡ğŸ‘‡ğŸ‘‡ğŸ‘‡ Horizontal Movement for Keyboard Input ğŸ‘‡ğŸ‘‡ğŸ‘‡ğŸ‘‡ #
			var direction = Input.get_axis("move_left", "move_right")
			# ğŸ‘‡ If direction is not = 0
			if direction:
				velocity.x = direction * speed
			else:
				# "speed/6" is the value of "delta"
				velocity.x = move_toward(velocity.x, 0, speed/6)
			# â˜ï¸ This takes value of velocity.x and set it to be 0 over a small period of time
		
	# ğŸ‘‡ Add this to make the player move
	move_and_slide()
	
	var margin = 20
	# ğŸ‘‡ To check if the player has reached right end of screen
	if global_position.x > viewport_size.x + margin:
		# ğŸ‘‡ if so, then 
		global_position.x = -margin
	# ğŸ‘‡ To check if the player has reached left end of screen
	if global_position.x < 0 - margin:
		global_position.x = viewport_size.x + margin

func jump():
	velocity.y = jump_velocity
	SoundFX.play("Jump")

# ğŸ‘‡ To detect when the player goes off the screen
func _on_visible_on_screen_notifier_2d_screen_exited() -> void:
	die()

# ğŸ‘‡ To disable collision shape of the player when the player is dead
func die():
	# ğŸ‘‡ If the player is still alive
	if !dead:
		dead = true
		# ğŸ‘‡ Alternative to [cshape.disabled = true] to avoid errors
		cshape.set_deferred("disabled", true)
		# ğŸ‘‡ Emitting the signal when the player dies
		died.emit()
		SoundFX.play("Fall")

# ğŸ‘‡ To use new skins that player must buy
func use_new_skin():
	# ğŸ‘‡ Changing the variable name to use red character's animation
	fall_anim_name = "fall_red"
	jump_anim_name = "jump_red"
	# ğŸ‘‡ Changing the texture of "Sprite2D" node to avoid 
	# ğŸ‘‡ begining frames with player's old skin.
	if sprite:
		sprite.texture = preload("res://assets/textures/character/Skin2Idle.png")
